File: crates/apollo_network/src/bin/broadcast_network_stress_test_node/run/args.py
import argparse


def add_shared_args_to_parser(parser: argparse.ArgumentParser):
    """
    Adds the arguments that are shared between the local and cluster deployment scripts
    """
    parser.add_argument(
        "--num-nodes", help="Number of nodes to run", type=int, default=3
    )
    parser.add_argument(
        "--verbosity",
        help="Verbosity level for logging (0: None, 1: ERROR, 2: WARN, 3: INFO, 4: DEBUG, 5..: TRACE)",
        type=int,
        default=2,
    )
    parser.add_argument(
        "--buffer-size",
        help="Buffer size to use by default.",
        type=int,
        default=100000,
    )
    parser.add_argument(
        "--message-size-bytes",
        help="Buffer size to use by default.",
        type=int,
        default=1 << 10,
    )
    parser.add_argument(
        "--heartbeat-millis",
        help="Number of milliseconds to wait between consecutive broadcasts.",
        type=int,
        default=1000,
    )
    parser.add_argument(
        "--mode",
        help="The mode to use for the stress test.",
        choices=["all", "one", "rr", "explore"],
        default="all",
    )
    parser.add_argument(
        "--network-protocol",
        help="The network protocol to use for communication.",
        choices=["gossipsub", "sqmr", "reversed-sqmr", "propeller"],
        default="gossipsub",
    )
    parser.add_argument(
        "--broadcaster",
        help="In mode `one`, which node ID should do the broadcasting, default is the last node.",
        type=int,
        default=None,
    )
    parser.add_argument(
        "--round-duration-seconds",
        help="Duration each node broadcasts before switching (in seconds) - for RoundRobin mode",
        type=int,
        default=3,
    )
    parser.add_argument(
        "--quic",
        help="Sets the multi-addresses to use UDP/QUIC instead of TCP",
        action="store_true",
        default=False,
    )
    parser.add_argument(
        "--explore-cool-down-duration-seconds",
        help="Cool down duration between configuration changes in seconds - for Explore mode",
        type=int,
        default=100,
    )
    parser.add_argument(
        "--explore-run-duration-seconds",
        help="Duration to run each configuration in seconds - for Explore mode",
        type=int,
        default=100,
    )
    parser.add_argument(
        "--explore-min-throughput-byte-per-seconds",
        help="Minimum throughput in bytes per second - for Explore mode",
        type=float,
        default=100 * (1 << 10),  # 100 KB/s
    )
    parser.add_argument(
        "--explore-min-message-size-bytes",
        help="Minimum message size in bytes - for Explore mode",
        type=int,
        default=1 << 10,  # 1 KB
    )
    parser.add_argument(
        "--timeout",
        help="The timeout in seconds for the node. Note than when running in a cluster the pod will be redeployed.",
        type=int,
        default=7200,
    )


def get_arguments(
    id: int | None,
    metric_port: int,
    p2p_port: int,
    bootstrap_nodes: list[str],
    args: argparse.Namespace,
) -> list[tuple[str, str]]:
    result = [
        ("--metric-port", str(metric_port)),
        ("--p2p-port", str(p2p_port)),
        ("--bootstrap", ",".join(bootstrap_nodes)),
        ("--timeout", str(args.timeout)),
        ("--verbosity", str(args.verbosity)),
        ("--buffer-size", str(args.buffer_size)),
        ("--message-size-bytes", str(args.message_size_bytes)),
        ("--heartbeat-millis", str(args.heartbeat_millis)),
        ("--mode", str(args.mode)),
        ("--network-protocol", str(args.network_protocol)),
        (
            "--broadcaster",
            (
                str(args.broadcaster)
                if args.broadcaster is not None
                else str(args.num_nodes - 1)
            ),
        ),
        ("--round-duration-seconds", str(args.round_duration_seconds)),
        (
            "--explore-cool-down-duration-seconds",
            str(args.explore_cool_down_duration_seconds),
        ),
        ("--explore-run-duration-seconds", str(args.explore_run_duration_seconds)),
        (
            "--explore-min-throughput-byte-per-seconds",
            str(args.explore_min_throughput_byte_per_seconds),
        ),
        (
            "--explore-min-message-size-bytes",
            str(args.explore_min_message_size_bytes),
        ),
        ("--num-nodes", str(args.num_nodes)),
    ]
    if id is not None:
        result.insert(0, ("--id", str(id)))
    return result


def get_env_vars(
    id: int | None,
    metric_port: int,
    p2p_port: int,
    bootstrap_nodes: list[str],
    args: argparse.Namespace,
) -> list[dict[str, str]]:
    arguments = get_arguments(
        id=id,
        metric_port=metric_port,
        p2p_port=p2p_port,
        bootstrap_nodes=bootstrap_nodes,
        args=args,
    )

    env_vars = []

    # Convert arguments to environment variables
    for name, value in arguments:
        env_name = name[2:].replace("-", "_").upper()
        env_vars.append({"name": env_name, "value": str(value)})

    # Add latency and throughput if provided
    for arg_name, env_name in [("latency", "LATENCY"), ("throughput", "THROUGHPUT")]:
        value = getattr(args, arg_name)
        if value is not None:
            env_vars.append({"name": env_name, "value": str(value)})

    return env_vars

File: crates/apollo_network/src/bin/broadcast_network_stress_test_node/args.rs
use std::fmt::Display;

use clap::{Parser, ValueEnum};

#[derive(Debug, Clone, ValueEnum, PartialEq, Eq)]
pub enum Mode {
    /// All nodes broadcast messages
    #[value(name = "all")]
    AllBroadcast,
    /// Only the node specified by --broadcaster broadcasts messages
    #[value(name = "one")]
    OneBroadcast,
    /// Nodes take turns broadcasting in round-robin fashion
    #[value(name = "rr")]
    RoundRobin,
    /// Only the node specified by --broadcaster broadcasts messages,
    /// Every explore_run_duration_seconds + explore_cool_down_duration_seconds seconds
    /// a new combination of MPS and message size is explored.
    /// Increases the throughput with each new trial.
    /// Configurations are filtered by minimum throughput and minimum message size.
    #[value(name = "explore")]
    Explore,
}

#[derive(Debug, Clone, ValueEnum, PartialEq, Eq)]
pub enum NetworkProtocol {
    /// Use gossipsub for broadcasting (default)
    #[value(name = "gossipsub")]
    Gossipsub,
    /// Use SQMR for point-to-point communication
    #[value(name = "sqmr")]
    Sqmr,
    /// Use Reversed SQMR where receivers initiate requests to broadcasters
    #[value(name = "reversed-sqmr")]
    ReveresedSqmr,
    /// Use Propeller for leader-based erasure-coded broadcasting
    #[value(name = "propeller")]
    Propeller,
}

impl Display for Mode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_possible_value().unwrap().get_name())
    }
}

impl Display for NetworkProtocol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_possible_value().unwrap().get_name())
    }
}

impl NetworkProtocol {
    /// Validates protocol-specific requirements for the given message size.
    /// Returns an error message if validation fails, or Ok(()) if validation passes.
    pub fn validate_message_size(&self, message_size_bytes: Option<usize>) -> Result<(), String> {
        match self {
            NetworkProtocol::Propeller => {
                if let Some(message_size) = message_size_bytes {
                    let default_config = libp2p::propeller::Config::default();
                    let shreds =
                        default_config.fec_coding_shreds() + default_config.fec_data_shreds();
                    if message_size % shreds != 0 {
                        return Err(format!(
                            "Propeller protocol requires message size to be a multiple of 64 \
                             bytes, got {}",
                            message_size
                        ));
                    }
                    let shred_size = message_size / shreds;
                    if !(shred_size).is_multiple_of(2) {
                        return Err(format!(
                            "Propeller protocol requires shred size to be a multiple of 2, got {}",
                            shred_size
                        ));
                    }
                }
                Ok(())
            }
            NetworkProtocol::Gossipsub | NetworkProtocol::Sqmr | NetworkProtocol::ReveresedSqmr => {
                // These protocols don't have specific message size requirements
                Ok(())
            }
        }
    }
}

#[derive(Parser, Debug, Clone)]
#[command(version, about, long_about = None)]
pub struct Args {
    /// ID for Prometheus logging
    #[arg(short, long, env)]
    pub id: u64,

    /// Total number of nodes in the network
    #[arg(long, env)]
    pub num_nodes: u64,

    /// The port to run the Prometheus metrics server on
    #[arg(long, env)]
    pub metric_port: u16,

    /// The port to run the P2P network on
    #[arg(short, env, long)]
    pub p2p_port: u16,

    /// The addresses of the bootstrap peers (can specify multiple)
    #[arg(long, env, value_delimiter = ',')]
    pub bootstrap: Vec<String>,

    /// Set the verbosity level of the logger, the higher the more verbose
    #[arg(short, long, env)]
    pub verbosity: u8,

    /// Buffer size for the broadcast topic
    #[arg(long, env)]
    pub buffer_size: usize,

    /// The mode to use for the stress test.
    #[arg(long, env)]
    pub mode: Mode,

    /// The network protocol to use for communication (default: gossipsub)
    #[arg(long, env)]
    pub network_protocol: NetworkProtocol,

    /// Which node ID should do the broadcasting - for OneBroadcast and Explore modes
    #[arg(long, env, required_if_eq_any([("mode", "one"), ("mode", "explore")]))]
    pub broadcaster: Option<u64>,

    /// Duration each node broadcasts before switching (in seconds) - for RoundRobin mode
    #[arg(long, env, required_if_eq("mode", "rr"))]
    pub round_duration_seconds: Option<u64>,

    /// Size of StressTestMessage
    #[arg(long, env, required_if_eq_any([("mode", "one"), ("mode", "all"), ("mode", "rr")]))]
    pub message_size_bytes: Option<usize>,

    /// The time to sleep between broadcasts of StressTestMessage in milliseconds
    #[arg(long, env, required_if_eq_any([("mode", "one"), ("mode", "all"), ("mode", "rr")]))]
    pub heartbeat_millis: Option<u64>,

    /// Cool down duration between configuration changes in seconds - for Explore mode
    #[arg(long, env, required_if_eq("mode", "explore"))]
    pub explore_cool_down_duration_seconds: Option<u64>,

    /// Duration to run each configuration in seconds - for Explore mode
    #[arg(long, env, required_if_eq("mode", "explore"))]
    pub explore_run_duration_seconds: Option<u64>,

    /// Minimum throughput in bytes per second - for Explore mode
    #[arg(long, env, required_if_eq("mode", "explore"))]
    pub explore_min_throughput_byte_per_seconds: Option<f64>,

    /// Minimum message size in bytes - for Explore mode
    #[arg(long, env, required_if_eq("mode", "explore"))]
    pub explore_min_message_size_bytes: Option<usize>,

    /// The timeout in seconds for the node.
    /// When the node runs for longer than this, it will be killed.
    #[arg(long, env)]
    pub timeout: u64,
}
